server/src/main/java/io/druid/client/CoordinatorServerView.java

```diff
@@ -0,0 +1,212 @@
+            return ServerView.CallbackAction.CONTINUE;
+          }
+
+          @Override
+          public ServerView.CallbackAction segmentViewInitialized()
+          {
+            initialized = true;
+            return ServerView.CallbackAction.CONTINUE;
+          }
+        }
+    );
+
+    baseView.registerServerCallback(
+        exec,
+        new ServerView.ServerCallback()
+        {
+          @Override
+          public ServerView.CallbackAction serverRemoved(DruidServer server)
+          {
+            removeServer(server);
+            return ServerView.CallbackAction.CONTINUE;
+          }
+        }
+    );
+  }
+
+  public boolean isInitialized()
+  {
+    return initialized;
+  }
+
+  public void clear()
+  {
+    synchronized (lock) {
+      timelines.clear();
+      selectors.clear();
+    }
+  }
+
+  private void removeServer(DruidServer server)
+  {
+    for (DataSegment segment : server.getSegments().values()) {
+      serverRemovedSegment(server.getMetadata(), segment);
+    }
+  }
+
+  private void serverAddedSegment(final DruidServerMetadata server, final DataSegment segment)
+  {
+    String segmentId = segment.getIdentifier();
+    synchronized (lock) {
+      log.debug("Adding segment[%s] for server[%s]", segment, server);
+
+      Set<DruidServerMetadata> servers = selectors.get(segmentId);
+      if (servers == null) {
+        servers = Sets.newHashSet();
+
+        VersionedIntervalTimeline<String, SegmentLoadInfo> timeline = timelines.get(segment.getDataSource());
+        if (timeline == null) {
+          timeline = new VersionedIntervalTimeline<>(Ordering.natural());
+          timelines.put(segment.getDataSource(), timeline);
+        }
+
+        timeline.add(
+            segment.getInterval(),
+            segment.getVersion(),
+            segment.getShardSpec().createChunk(new SegmentLoadInfo(segment, servers))
+        );
+        selectors.put(segmentId, servers);
+      }
+      servers.add(server);
+    }
+  }
+
+  private void serverRemovedSegment(DruidServerMetadata server, DataSegment segment)
+  {
+    String segmentId = segment.getIdentifier();
+    final Set<DruidServerMetadata> servers;
+
+    synchronized (lock) {
+      log.debug("Removing segment[%s] from server[%s].", segmentId, server);
+
+      servers = selectors.get(segmentId);
+      if (servers == null) {
+        log.warn("Told to remove non-existant segment[%s]", segmentId);
+        return;
+      }
+      servers.remove(server);
+      if (servers.isEmpty()) {
+        VersionedIntervalTimeline<String, SegmentLoadInfo> timeline = timelines.get(segment.getDataSource());
+        selectors.remove(segmentId);
+
+        final PartitionChunk<SegmentLoadInfo> removedPartition = timeline.remove(
+            segment.getInterval(), segment.getVersion(), segment.getShardSpec().createChunk(
+                new SegmentLoadInfo(
+                    segment,
+                    Sets
```

Comment:
ok, I see, you are modifying it. Those modifications should be made thread safe then, as the SegmentLoadInfos are getting escaped outside of this object.